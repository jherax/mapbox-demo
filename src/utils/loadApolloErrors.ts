import {loadDevMessages, loadErrorMessages} from '@apollo/client/dev';

export default function loadApolloErrors() {
  if (!['prod', 'production'].includes(process.env.NODE_ENV)) {
    loadDevMessages();
    loadErrorMessages();
  }
}
