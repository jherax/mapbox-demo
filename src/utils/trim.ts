export default function trim(text?: string | null): string {
  if (!text) {
    return '';
  }
  return text.trim();
}
