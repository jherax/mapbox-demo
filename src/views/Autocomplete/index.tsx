import {useEffect, useRef, useState} from 'react';

import Countries from '../../components/Countries';
import buildKey from '../../utils/buildKey';
import ResolveView from '../ResolveView';
import {useLazyCitiesByCountry} from './services/hooks';
import {
  MainWrapper,
  ResultsContainer,
  ResultsItem,
  TagsContainer,
  TagsItem,
} from './styled';

const ucase = (text: string) => text.toUpperCase();
const removeSpace = (text: string) => text.toLowerCase().replaceAll(' ', '-');
const restoreSpace = (text: string) => text.replaceAll('-', ' ');
const ALL_CITIES: Record<string, CityResults[]> = Object.create(null);

type CityResults = City & {
  hidden?: boolean;
};

export type AutocompleteProps = Readonly<{
  maxWidth: string;
}>;

export default function Autocomplete(props: AutocompleteProps) {
  const [disabledInput, setDisabledInput] = useState(true);
  const {getCities, error, data} = useLazyCitiesByCountry();

  const onChangeCountry = (country: string) => {
    setDisabledInput(!country);
    cleanInputText(inputRef);
    setShowResults(false);
    if (country) {
      getCities({variables: {country}});
    }
  };

  useEffect(() => {
    setDisabledInput(!data);
    const cities = data?.getCitiesByCountry?.result;
    if (cities != null) {
      const {country} = cities[0];
      if (ALL_CITIES[country] == null) {
        ALL_CITIES[country] = getUniqueCities(cities);
      }
      ALL_CITIES['current'] = ALL_CITIES[country];
    }
  }, [data]);

  const [tags, setTags] = useState<string[]>([]);
  const [results, setResults] = useState<CityResults[]>([]);
  const [showResults, setShowResults] = useState(false);
  const inputRef = useRef<HTMLInputElement>(null);

  const handleKeyUp: React.KeyboardEventHandler<HTMLInputElement> = e => {
    const currentText = (e.target as HTMLInputElement).value;
    if (e.code === 'Escape') {
      cleanInputText(inputRef);
      setShowResults(false);
    } else if (e.code === 'Enter') {
      onClickSuggestion(currentText);
    } else if (currentText) {
      const prefix = ucase(currentText);
      // search criteria: match cities starting with ...
      const matchedCities = ALL_CITIES['current'].filter(city => {
        return !city.hidden && ucase(city.name).startsWith(prefix);
      });
      setResults(matchedCities);
      if (matchedCities.length && !showResults) {
        setShowResults(true);
      }
    }
  };

  const onClickSuggestion = (suggestion: string) => {
    const tag = removeSpace(suggestion);
    if (tag && tags.indexOf(tag) === -1) {
      setTags([...tags, tag]);
    }
    const selectedCity = ALL_CITIES['current'].find(
      city => ucase(city.name) === ucase(suggestion),
    );
    if (selectedCity != null) {
      selectedCity.hidden = true;
    }
    cleanInputText(inputRef);
    setShowResults(false);
  };

  const onClickTag = (tag: string) => {
    setTags(tags.filter(t => t !== tag));
    const cityTag = restoreSpace(ucase(tag));
    const selectedCity = ALL_CITIES['current'].find(
      city => ucase(city.name) === cityTag,
    );
    if (selectedCity != null) {
      selectedCity.hidden = false;
    }
    setShowResults(false);
  };

  // logger.info({loading, data, showResults, disabledInput})
  return (
    <MainWrapper $maxWidth={props.maxWidth}>
      <fieldset>
        <legend>Cities around the world</legend>
        <Countries onChangeCountry={onChangeCountry}></Countries>
        <ResolveView error={error}>
          <TagsContainer className='autocomplete_container'>
            {tags.map((tag, i) => {
              const key = `tag-${buildKey()}-${i}`;
              return (
                <TagsItem
                  key={key}
                  value={tag}
                  onClick={onClickTag}
                  className='autocomplete_tags_item'
                ></TagsItem>
              );
            })}
            <input
              type='text'
              name='cities_searchbox'
              className='autocomplete_input'
              placeholder='Cities...'
              ref={inputRef}
              disabled={disabledInput}
              onKeyUp={handleKeyUp}
            />
          </TagsContainer>
          <ResultsContainer
            $show={+showResults}
            className='autocomplete_results'
          >
            {results.map((result, i) => {
              const key = `result-${buildKey()}-${i}`;
              return (
                <ResultsItem
                  key={key}
                  value={result.name}
                  onClick={onClickSuggestion}
                  className='autocomplete_results_item'
                ></ResultsItem>
              );
            })}
          </ResultsContainer>
        </ResolveView>
      </fieldset>
    </MainWrapper>
  );
}

// ---------------------------------

function getUniqueCities(cities: CityResults[]): CityResults[] {
  const uniqueCities: CityResults[] = [];
  const total = cities.length;
  for (let i = 0; i < total; i += 1) {
    const city = cities[i];
    if (!uniqueCities.some(c => c.name === city.name)) {
      // this fixes the error: object is not extensible
      uniqueCities.push({...city});
    }
  }
  return uniqueCities;
}

function cleanInputText(ref: React.RefObject<HTMLInputElement>) {
  if (ref.current != null) {
    ref.current.value = '';
  }
}
