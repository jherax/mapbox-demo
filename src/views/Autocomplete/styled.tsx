import {styled} from 'styled-components';

const $marginLeft = '5px';

export const MainWrapper = styled.section<{$maxWidth: string}>`
  color: #444;
  fieldset {
    max-width: ${props => props.$maxWidth};
    position: relative;

    select,
    input {
      margin-left: ${$marginLeft};
      color: #444;
      border: none;
      &:focus,
      &:active {
        outline: none;
      }
    }

    section {
      margin: 8px;
      padding: 8px;
      &:not(.countries_container) {
        border: solid 1px #444;
      }
    }

    section.countries_container {
      label {
        margin-right: 5px;
        font-size: 0.9em;
      }
      select.countries_dropdown {
        cursor: pointer;
      }
    }

    section.autocomplete_container {
      display: flex;
      flex-flow: row wrap;
      border-style: dashed;

      input.autocomplete_input {
        flex-grow: 1;
      }
      label.autocomplete_tags_item {
        margin: ${$marginLeft};
        padding: 4px 7px;
        box-sizing: border-box;
        border: solid 1px #5e8da0;
        background-color: rgba(142, 175, 189, 0.1);
        font-size: 0.8em;
        &:hover {
          border: solid 1px #a05e6c;
          background-color: rgba(160, 94, 108, 0.1);
        }
      }
    }

    section.autocomplete_results {
      /* position: absolute; */
      z-index: 1;
      margin-top: 3px;
      background: white;
      color: #666;
      font-size: 0.9em;

      div.autocomplete_results_item {
        cursor: pointer;
        padding: 4px;
        &:hover {
          background-color: rgba(207, 200, 174, 0.4);
        }
      }
    }
  }
`;

// ---------------------------------

export const TagsContainer = styled.section``;
const TagElement = styled.label``;
export type TagsItemProps = Readonly<{
  value: string;
  className: string;
  onClick: (tag: string) => void;
}>;

export function TagsItem(props: TagsItemProps) {
  const handleClick = () => props.onClick(props.value);
  return (
    <TagElement className={props.className} onClick={handleClick}>
      {props.value}
    </TagElement>
  );
}

// ---------------------------------

export const ResultsContainer = styled.section<{$show: number}>`
  display: ${({$show}) => ($show ? 'block' : 'none')};
`;

const Suggestion = styled.div``;
export type ResultsItemProps = Readonly<{
  value: string;
  className: string;
  onClick: (value: string) => void;
}>;

export function ResultsItem(props: ResultsItemProps) {
  const handleClick = () => props.onClick(props.value);
  return (
    <Suggestion
      className={props.className}
      onClick={handleClick}
      role='listitem'
    >
      {props.value}
    </Suggestion>
  );
}
