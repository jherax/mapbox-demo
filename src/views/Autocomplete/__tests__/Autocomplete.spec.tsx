import {MockedProvider} from '@apollo/client/testing';
import {
  cleanup,
  fireEvent,
  queryByAttribute,
  render,
  screen,
  waitFor,
} from '@testing-library/react';

import loadApolloErrors from '../../../utils/loadApolloErrors';
import trim from '../../../utils/trim';
import Autocomplete from '..';
import {
  emptyCitiesByCountryMock,
  errorCitiesByCountryMock,
  successCitiesByCountryMock,
  variablesMock,
} from '../__mocks__/apolloAutocomplete';

loadApolloErrors();
const getByName = queryByAttribute.bind(null, 'name');
const getBySelector = (parent: HTMLElement, selector: string) =>
  // eslint-disable-next-line testing-library/no-node-access
  parent.querySelector(selector);

/**
 * @see https://jestjs.io/docs/snapshot-testing
 * @see https://testing-library.com/docs/react-testing-library/setup
 */
describe('Testing <Autocomplete />', () => {
  afterEach(() => cleanup());

  it('should get cities by country, trigger the autocomplete and select the first suggestion', async () => {
    const {asFragment, container} = render(
      <MockedProvider mocks={successCitiesByCountryMock} addTypename={false}>
        <Autocomplete maxWidth='500px' />
      </MockedProvider>,
    );

    // Initial state.
    const dropdown = getByName(container, 'countries') as HTMLSelectElement;
    const input = getByName(container, 'cities_searchbox') as HTMLInputElement;
    const results = getBySelector(container, '.autocomplete_results');
    expect(results?.childNodes).toHaveLength(0);
    expect(input).toBeDisabled();

    fireEvent.change(dropdown, {target: {value: variablesMock.country}});
    // Wait to get cities by selected country.
    expect(dropdown.value).toBe('Colombia');
    await waitFor(async () => {
      expect(input).not.toBeDisabled();
    });

    // Set the value of the searchbox input.
    fireEvent.change(input, {target: {value: 'a'}});
    fireEvent.keyUp(input, {code: 'KeyA', key: 'a', keyCode: 65});

    // Wait for search results for cities starting with "A".
    // Each suggestion item has the "listitem" role.
    const suggestions = await screen.findAllByRole('listitem');
    expect(suggestions).toHaveLength(2);
    expect(results?.childNodes).toHaveLength(2);

    // Adds the first suggested City to the Tags list.
    const firstSuggestion = suggestions[0];
    fireEvent.click(firstSuggestion);
    await waitFor(() => {
      const tag = getBySelector(container, '.autocomplete_tags_item');
      expect(tag).toBeInTheDocument();
    });

    const firstTag = getBySelector(container, '.autocomplete_tags_item');
    const selectedText = trim(firstSuggestion.textContent).toLowerCase();
    expect(firstTag?.textContent).toMatch(selectedText);
    expect(asFragment()).toMatchSnapshot();
  });

  // ---------------------------------

  it('should add a tag when pressing ENTER', async () => {
    const {asFragment, container} = render(
      <MockedProvider mocks={successCitiesByCountryMock} addTypename={false}>
        <Autocomplete maxWidth='500px' />
      </MockedProvider>,
    );

    // Initial state.
    const dropdown = getByName(container, 'countries') as HTMLSelectElement;
    const input = getByName(container, 'cities_searchbox') as HTMLInputElement;
    expect(input).toBeDisabled();

    // Selects a country to enable the searchbox input
    fireEvent.change(dropdown, {target: {value: variablesMock.country}});
    await waitFor(async () => {
      expect(input).not.toBeDisabled();
    });

    // Set the value in the searchbox input and press ENTER.
    fireEvent.change(input, {target: {value: 'Svalbard'}});
    fireEvent.keyUp(input, {code: 'Enter', key: 'Enter', keyCode: 13});
    // Wait for the tag to be added
    await waitFor(() => {
      const tag = getBySelector(container, '.autocomplete_tags_item');
      expect(tag).toBeInTheDocument();
    });

    const tag = getBySelector(container, '.autocomplete_tags_item');
    expect(tag?.textContent).toMatch('svalbard');
    expect(asFragment()).toMatchSnapshot();
  });

  // ---------------------------------

  it('should render correctly when there are no fetched data', async () => {
    const {asFragment, container} = render(
      <MockedProvider mocks={emptyCitiesByCountryMock} addTypename={false}>
        <Autocomplete maxWidth='500px' />
      </MockedProvider>,
    );

    expect(getByName(container, 'cities_searchbox')).toBeDisabled();
    expect(asFragment()).toMatchSnapshot();
  });

  // ---------------------------------

  it('should render the error view', async () => {
    const {asFragment, container} = render(
      <MockedProvider mocks={errorCitiesByCountryMock} addTypename={false}>
        <Autocomplete maxWidth='500px' />
      </MockedProvider>,
    );

    // triggers the getCities() lazy-query
    const dropdown = getByName(container, 'countries') as HTMLSelectElement;
    fireEvent.change(dropdown, {target: {value: variablesMock.country}});

    const errorMessage = 'Unknown server error when fetching data';
    expect(await screen.findByText(errorMessage)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
