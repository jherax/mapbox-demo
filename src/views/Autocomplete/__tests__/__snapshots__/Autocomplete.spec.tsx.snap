// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <Autocomplete /> should add a tag when pressing ENTER 1`] = `
<DocumentFragment>
  <section
    class="sc-guDLey fUHaaq"
  >
    <fieldset>
      <legend>
        Cities around the world
      </legend>
      <section
        class="countries_container"
      >
        <label
          for=":r1:"
        >
          Search by country:
        </label>
        <select
          class="countries_dropdown"
          id=":r1:"
          name="countries"
        >
          <option
            value=""
          >
            select...
          </option>
          <option
            value="Colombia"
          >
            Colombia
          </option>
          <option
            value="United States"
          >
            United States
          </option>
        </select>
      </section>
      <section
        class="sc-dmyCSP autocomplete_container"
      >
        <label
          class="sc-hLQSwg autocomplete_tags_item"
        >
          svalbard
        </label>
        <input
          class="autocomplete_input"
          name="cities_searchbox"
          placeholder="Cities..."
          type="text"
        />
      </section>
      <section
        class="sc-eDLKkx dTMqXE autocomplete_results"
      />
    </fieldset>
  </section>
</DocumentFragment>
`;

exports[`Testing <Autocomplete /> should get cities by country, trigger the autocomplete and select the first suggestion 1`] = `
<DocumentFragment>
  <section
    class="sc-guDLey fUHaaq"
  >
    <fieldset>
      <legend>
        Cities around the world
      </legend>
      <section
        class="countries_container"
      >
        <label
          for=":r0:"
        >
          Search by country:
        </label>
        <select
          class="countries_dropdown"
          id=":r0:"
          name="countries"
        >
          <option
            value=""
          >
            select...
          </option>
          <option
            value="Colombia"
          >
            Colombia
          </option>
          <option
            value="United States"
          >
            United States
          </option>
        </select>
      </section>
      <section
        class="sc-dmyCSP autocomplete_container"
      >
        <label
          class="sc-hLQSwg autocomplete_tags_item"
        >
          arinsal
        </label>
        <input
          class="autocomplete_input"
          name="cities_searchbox"
          placeholder="Cities..."
          type="text"
        />
      </section>
      <section
        class="sc-eDLKkx dTMqXE autocomplete_results"
      >
        <div
          class="sc-jTQCzO autocomplete_results_item"
          role="listitem"
        >
          Arinsal
        </div>
        <div
          class="sc-jTQCzO autocomplete_results_item"
          role="listitem"
        >
          Andorra la Vella
        </div>
      </section>
    </fieldset>
  </section>
</DocumentFragment>
`;

exports[`Testing <Autocomplete /> should render correctly when there are no fetched data 1`] = `
<DocumentFragment>
  <section
    class="sc-guDLey fUHaaq"
  >
    <fieldset>
      <legend>
        Cities around the world
      </legend>
      <section
        class="countries_container"
      >
        <label
          for=":r2:"
        >
          Search by country:
        </label>
        <select
          class="countries_dropdown"
          id=":r2:"
          name="countries"
        >
          <option
            value=""
          >
            select...
          </option>
          <option
            value="Colombia"
          >
            Colombia
          </option>
          <option
            value="United States"
          >
            United States
          </option>
        </select>
      </section>
      <section
        class="sc-dmyCSP autocomplete_container"
      >
        <input
          class="autocomplete_input"
          disabled=""
          name="cities_searchbox"
          placeholder="Cities..."
          type="text"
        />
      </section>
      <section
        class="sc-eDLKkx dTMqXE autocomplete_results"
      />
    </fieldset>
  </section>
</DocumentFragment>
`;

exports[`Testing <Autocomplete /> should render the error view 1`] = `
<DocumentFragment>
  <section
    class="sc-guDLey fUHaaq"
  >
    <fieldset>
      <legend>
        Cities around the world
      </legend>
      <section
        class="countries_container"
      >
        <label
          for=":r3:"
        >
          Search by country:
        </label>
        <select
          class="countries_dropdown"
          id=":r3:"
          name="countries"
        >
          <option
            value=""
          >
            select...
          </option>
          <option
            value="Colombia"
          >
            Colombia
          </option>
          <option
            value="United States"
          >
            United States
          </option>
        </select>
      </section>
      <section
        class="sc-beySPh duqtwh ResolveView_error"
      >
        <pre>
          Unknown server error when fetching data
        </pre>
      </section>
    </fieldset>
  </section>
</DocumentFragment>
`;
